### DISCUSSION TITLE
Open file in Adobe Reader

### DISCUSSION THREAD

**Discussion by dumb0x** [sumatrapdfreader/sumatrapdf#4136] [Q&A]
On updating to v.3.5.2, I notice that the function to Open file in Adobe Reader (Alt-F,A) is gone, but I don't find any note about that in release notes. When did that happen? What was the motivation for the change?

**Comment by GitHubRulesOK** [comment]
It should still be there (it is in pre-release), but detection is based on correct relevant registry value. 

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/519799b3-9c30-430b-8ed8-7503a23f1d73)

**Reply by dumb0x** [reply]
Interesting. You show **_three_** items that are not in my v.3.5.2 menu:
- Open Directory in Explorer
- Open in Adobe Reader
- Send by E-mail...

I have _**never**_ seen the first or third of those, which sound (especially the first) very useful.  The second is the item I asked about, but now I must ask about all three: What is needed to have these commands in the menu?  Is there documentation on this issue?

**Reply by GitHubRulesOK** [reply]
Again Send by email is an old entry and just to confuse you my 3.5.2 has 5 !  

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/3f9f35bb-a02b-4697-914d-a064f738eee9)
 
The extra 2 are custom entries one runs a notepad goto line number for PDFLaTeX edits the other was a personal related  test of recalling SumatraPDF. So the 3 under discussion are standard 3.5.2 on this Windows 10 which has no mail ! but does have Acrobat PDF reader installed, so the option would only be valid for PDF files, although Acrobat can handle more than one file type.
```
ExternalViewers [
	[
		CommandLine = "c:\windows\System32\cmd.exe" /s /k echo %1-Page%p.txt
		Name = Notepad Scrip&t
		Filter = *.*
	]
	[
		CommandLine = "c:\windows\System32\cmd.exe" /k echo "C:\Users\God\AppData\Local\SumatraPDF\SumatraPDF.exe" "%1" "-page" %p
		Name = SumatraPDF
		Filter = *.pdf
	]
]
```

**Reply by dumb0x** [reply]
Well, I give up, Krzysztof [is that you?].  

I found "ExternalViewers" in AppData\Local\SumatraPDF\SumatraPDF-settings.txt, but nothing instantiated there, and certainly no "CommandLine" invocations anywhere in the file.  So I presumed you were talking about Windows registry entries -- which would be too bad, because I'm not steeped in the registry, and have only the clumsy regedit to manage it.  
Still, I looked, and found maybe hundreds of "SumatraPDF keys and values, but nothing like the commands you posted. So where are they, and how do I get my questions answered?  I just want to be able to exploit SumatraPDF's functions - at least to get back to the "Open file in Adobe Reader" that I had before updating.

**Reply by GitHubRulesOK** [reply]
@dumb0x 

No I an not KJK (Krzysztof) my initials are KJ... (lots of family traditional names! I am K and J was my Dad and granddad etc)

The extra examples I show in external are my own text additions, so as shown below recently testing export as PNG.
However if Acrobat is not AUTO detected you can add your own.

Look for Acro Reader it should be `"C:\Program Files\Adobe\Acrobat Reader DC\Reader\AcroRd32.exe"`

So have added another entry the /n opens acrobat with new window and the shortcut is Alt + F + R
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/08565613-2ebc-44f9-8a7b-f36330aa984a)

```
ExternalViewers [
	[
		CommandLine = "C:\Program Files\Adobe\Acrobat Reader DC\Reader\AcroRd32.exe" /n "%1"
		Name = External Acrobat &Reader
		Filter = *.pdf
	]
]
```

**Comment by kjk** [comment]
@dumb0x You can see our Acrobat at:
https://github.com/sumatrapdfreader/sumatrapdf/blob/master/src/ExternalViewers.cpp#L79
https://github.com/sumatrapdfreader/sumatrapdf/blob/master/src/ExternalViewers.cpp#L213

It's possible not every install of Adobe Reader writes the registry info the way we read it.

If you can tell me path of Adobe Reader executable on your computer, I can add a check for it.

**Reply by dumb0x** [reply]
Thanks for clarifying that I should look in ExternalViewers.cpp, edit as needed and recompile. That is more than can think about for the next week, because my Adobe reader is at:

"C:\Program Files\Adobe\Acrobat DC\Acrobat\Acrobat.exe", _not_ 
"C:\Program Files (x86)\Adobe\Acrobat Reader DC\Reader\AcroRd32.exe" (ExternalViewers.cpp L#83)

- and I don't know why that is. (I probably knew at one point and have forgotten - there are other AcroRd32 files on the computer, but not AcroRd32.exe - so will need to think it through before making code change.)  I also saw what you are doing at lines 213_ff_, but again mine is different, and I should study the source thoroughly to see the logic.

I any case, I am now also interested in the "Open Directory in Explorer" and "Send by E-mail..." entries which I learned about in this thread (never saw those entries before), which are _not_ in ExternalViewers.cpp, so I'm afraid my question has only expanded, and I will need to take it up again later.  Thanks for your reply.

**Reply by GitHubRulesOK** [reply]
Acrobat.exe should be the FULL writer not the lighter reader and often the 2 are different locations and capabilities with different registry settings. The detector is primarily looking for most common "user"s Reader version.

Open directory in explorer is I think a more recent addition so perhaps your installation is older?

But email has been there for "ever" when available. However it may not see any such system ability, I have never used it more than a couple of times in a decade, so would not miss it.
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/7e9b6a06-4534-48a3-bee9-48ce70fb61cc)

**Comment by kjk** [comment]
@dumb0x I've added detecting your Adobe Reader path in https://github.com/sumatrapdfreader/sumatrapdf/commit/4e0d7592643dbaaedcaa5b49a57cf29ebdb675eb

Should show up in pre-release 16009 or later. Let me know if it doesn't work.

`Open Directory in Explorer` is in `ExternalViewers.cpp`. It's `CmdOpenWithExplorer` entry.

`Send as email` is `CanSendAsEmailAttachment` in the code.

**Comment by dumb0x** [comment]
Thanks. That's a lot more than I expected. Still, I want to return to this later for understanding -- for example, something as basic as how the menu string "Open Directory in Explorer" connects with CmdOpenWithExplorer in ExternalViewers.cpp -- to see how these capabilities are added.
